### PR TITLE
Support multiple env yaml specs

### PR DESCRIPTION
**Pull request by jchorl**
Tested:

testmultiyml/env1.yml 
```yaml
channels:
  - conda-forge
dependencies:
  - python >=3.12
```

testmultiyml/env2.yml 
```yaml
channels:
  - conda-forge
dependencies:
  - pandas >=2.1.1
```

```
(mamba) mambauser@6eb769d94efc:/work$ ./build/micromamba/micromamba create -n multiymltest -f testmultiyml/env1.yml -f testmultiyml/env2.yml 
(mamba) mambauser@6eb769d94efc:/work$     micromamba activate multiymltest
(multiymltest) mambauser@6eb769d94efc:/work$ python3
Python 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
```

Not sure if there's other edge cases worth considering or test cases that would be helpful.

### CODE REVIEW

**Comment by jonashaag** [comment]
Thanks for working on this!

This definitely needs more tests and coverage of edge cases.

Some things that come to mind, not an exhaustive list:

- conflicting specs
- conflicting channels (defaults vs nodefaults)
- conflicting channel priorities 
- missing entries in one of the files
- passing non env.yml files as one of the files (should probably error)

**Comment by jchorl** [comment]
> Thanks for working on this!
> 
> This definitely needs more tests and coverage of edge cases.
> 
> Some things that come to mind, not an exhaustive list:
> 
>     * conflicting specs
> 
>     * conflicting channels (defaults vs nodefaults)
> 
>     * conflicting channel priorities
> 
>     * missing entries in one of the files
> 
>     * passing non env.yml files as one of the files (should probably error)

Thanks. These are great suggestions. I found a bit more time today to work on this.

General thought: It would be good to understand what new testing you feel is required just for multi-yaml-spec support. Happy to add more if you feel it makes sense. Thanks for the help thus far!


Re specific tests:

First, I modified the multi-spec test to properly handle multiple yamls.

> conflicting channels (defaults vs nodefaults)

I did write a test for this (removed after): `test_multiple_yaml_specs_conflicting_defaults`.

I ran this through micromamba:

```
micromamba install -n base pandas -c nodefaults -c defaults
```

This actually works. This makes me think the check for conflicting defaults belongs at a lower layer, not multi-yaml-spec parsing. Thoughts?

> conflicting channel priorities

Same as above.

I ran this through micromamba:
```
(base) mambauser@e7e89efc8d2c:/tmp$ micromamba install -n base pandas -c conda-forge -c bioconda -c bioconda -c conda-forge
```

It succeeded. This makes me think any channel conflict detection should be lower down. The problem feels not multi-yaml-spec-specific, as in, today you can pass `-c chan1 -c chan2` and have your yaml spec specify `channels: [chan2, chan1]`, and it won't error (as best I can tell).

Would you be open to punting this check to a future PR that solves conflicting channels more holistically?

> missing entries in one of the files

I wasn't sure what you meant here. I added `test_multiple_yaml_specs_only_one_has_channels` whereby only one yaml specifies the channels. The result is the union of all channels (so, only the one channel from the one yaml) but all the deps get concatted. This is expected/desirable to me. Did you mean to test for something else?

> passing non env.yml files as one of the files (should probably error)

done! added a test too.

File: micromamba/tests/test_create.py [hunk 1]
```diff
@@ -330,15 +330,64 @@ def test_multiple_spec_files(tmp_home, tmp_root_prefix, tmp_path, type):
...
+    assert "found multiple spec file types" in info.value.stderr.decode()
+
+
+@pytest.mark.parametrize("shared_pkgs_dirs", [True], indirect=True)
+def test_multiple_yaml_specs_only_one_has_channels(tmp_home, tmp_root_prefix, tmp_path):
+    env_prefix = tmp_path / "myenv"
+
+    cmd = ["-p", env_prefix]
+
+    spec_file_1 = tmp_path / f"env1.yaml"
```
Comment (jonashaag) [inline micromamba/tests/test_create.py:369]: See above

**Comment by jonashaag** [comment]
Looks pretty good already. What happens if you specify conflicting specs in the files? eg. python=3.11 and python=3.10. Is the error message helpful?

File: micromamba/tests/test_create.py [hunk 2]
```diff
@@ -330,15 +330,64 @@ def test_multiple_spec_files(tmp_home, tmp_root_prefix, tmp_path, type):
...
+@pytest.mark.parametrize("shared_pkgs_dirs", [True], indirect=True)
+def test_multiple_spec_files_different_types(tmp_home, tmp_root_prefix, tmp_path):
+    env_prefix = tmp_path / "myenv"
+
+    cmd = ["-p", env_prefix]
+
+    spec_file_1 = tmp_path / f"env1.yaml"
+    file_content = ["dependencies: [xtensor]"]
+    with open(spec_file_1, "w") as f:
+        f.write("\n".join(file_content))
```
[the remark this answers is not in the document]
Comment (jchorl) [reply micromamba/tests/test_create.py:349]: Huh - TIL `pathlib.Path` has a `write_text` method. Neat!

**Comment by jchorl** [comment]
I think it should be ready for another look!

**Review by jonashaag — APPROVED** [review]
Nice thank you!

**Review by AntoinePrv — APPROVED** [review]

**Comment by AntoinePrv** [comment]
Thank you for your patience @jchorl
